# Predicting Employee Attrition using Supervised Learning
## Task 1 - Data Understanding and Exploration

This notebook uses only the provided IBM HR Analytics Employee Attrition CSV file. It performs exploration only; no preprocessing, train-test split, or model training is performed in this task.

### Task 1 objectives

- Load and inspect the dataset
- Report dataset dimensions and feature types
- Examine the target class distribution
- Check missing values and duplicate rows
- Produce summary statistics
- Derive meaningful observations using the actual dataset

In [1]:
# Import libraries required for Task 1.
from pathlib import Path

import pandas as pd

# No random process is used in Task 1, so random_state is not needed.
DATA_FILE = Path('WA_Fn-UseC_-HR-Employee-Attrition.csv')
TARGET_COLUMN = 'Attrition'

if not DATA_FILE.exists():
    raise FileNotFoundError(
        f'Dataset file not found: {DATA_FILE.resolve()}\n'
        'Place the CSV file in the same folder as this notebook.'
    )

df = pd.read_csv(DATA_FILE)

if TARGET_COLUMN not in df.columns:
    raise ValueError(
        f"Expected target column '{TARGET_COLUMN}' was not found.\n"
        f'Available columns: {df.columns.tolist()}'
    )

print('Dataset loaded successfully.')
print(f'File name: {DATA_FILE.name}')
print(f'Dataset shape (rows, columns): {df.shape}')

Dataset loaded successfully.
File name: WA_Fn-UseC_-HR-Employee-Attrition.csv
Dataset shape (rows, columns): (1470, 35)


In [2]:
# 1. Dataset dimensions and initial inspection
print('=' * 70)
print('1. DATASET DIMENSIONS AND PREVIEW')
print('=' * 70)
print(f'Number of employees (rows): {df.shape[0]}')
print(f'Number of columns: {df.shape[1]}')
print(f'Number of predictor features: {df.shape[1] - 1}')

print('\nFirst five rows:')
print(df.head().to_string())

print('\nColumn names:')
print(df.columns.tolist())

1. DATASET DIMENSIONS AND PREVIEW
Number of employees (rows): 1470
Number of columns: 35
Number of predictor features: 34

First five rows:
   Age Attrition     BusinessTravel  DailyRate              Department  DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel                JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  StandardHours  StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager
0   41       Yes      Travel_Rarely       1102                   Sales                 1          2  Life Sciences              1               1                        2  Female          94               3         2        Sales Executive                4        Single           5993 

In [3]:
# 2. Feature types
print('=' * 70)
print('2. FEATURE TYPES')
print('=' * 70)

feature_information = pd.DataFrame({
    'Data Type': df.dtypes.astype(str),
    'Non-Null Values': df.notna().sum(),
    'Unique Values': df.nunique(dropna=True)
})

print(feature_information.to_string())

numeric_features = df.drop(columns=TARGET_COLUMN).select_dtypes(include='number').columns.tolist()
categorical_features = df.drop(columns=TARGET_COLUMN).select_dtypes(
    include=['object', 'string', 'category', 'bool']
).columns.tolist()

print(f'Numerical predictor features ({len(numeric_features)}):')
print(numeric_features)
print(f'\nCategorical predictor features ({len(categorical_features)}):')
print(categorical_features)
print(f'\nTarget column data type: {df[TARGET_COLUMN].dtype}')

2. FEATURE TYPES
                         Data Type  Non-Null Values  Unique Values
Age                          int64             1470             43
Attrition                   object             1470              2
BusinessTravel              object             1470              3
DailyRate                    int64             1470            886
Department                  object             1470              3
DistanceFromHome             int64             1470             29
Education                    int64             1470              5
EducationField              object             1470              6
EmployeeCount                int64             1470              1
EmployeeNumber               int64             1470           1470
EnvironmentSatisfaction      int64             1470              4
Gender                      object             1470              2
HourlyRate                   int64             1470             71
JobInvolvement               int64           

In [4]:
# 3. Target class distribution
print('=' * 70)
print('3. TARGET CLASS DISTRIBUTION')
print('=' * 70)

target_counts = df[TARGET_COLUMN].value_counts()
target_percentages = (df[TARGET_COLUMN].value_counts(normalize=True) * 100).round(2)

target_distribution = pd.DataFrame({
    'Employee Count': target_counts,
    'Percentage (%)': target_percentages
})
print(target_distribution.to_string())

3. TARGET CLASS DISTRIBUTION
           Employee Count  Percentage (%)
Attrition                                
No                   1233           83.88
Yes                   237           16.12


In [5]:
# 4. Missing-value and duplicate-row checks
print('=' * 70)
print('4. DATA QUALITY CHECKS')
print('=' * 70)

missing_values = pd.DataFrame({
    'Missing Values': df.isna().sum(),
    'Missing Percentage (%)': (df.isna().mean() * 100).round(2)
}).sort_values('Missing Values', ascending=False)

print('Missing values for every column:')
print(missing_values.to_string())

columns_with_missing_values = missing_values[missing_values['Missing Values'] > 0]
if columns_with_missing_values.empty:
    print('Result: No missing values were found in the dataset.')
else:
    print('Columns containing missing values:')
    print(columns_with_missing_values.to_string())

duplicate_rows = df.duplicated().sum()
print(f'\nNumber of fully duplicated rows: {duplicate_rows}')

4. DATA QUALITY CHECKS
Missing values for every column:
                          Missing Values  Missing Percentage (%)
Age                                    0                     0.0
Attrition                              0                     0.0
BusinessTravel                         0                     0.0
DailyRate                              0                     0.0
Department                             0                     0.0
DistanceFromHome                       0                     0.0
Education                              0                     0.0
EducationField                         0                     0.0
EmployeeCount                          0                     0.0
EmployeeNumber                         0                     0.0
EnvironmentSatisfaction                0                     0.0
Gender                                 0                     0.0
HourlyRate                             0                     0.0
JobInvolvement                    

In [6]:
# 5. Summary statistics
print('=' * 70)
print('5. SUMMARY STATISTICS')
print('=' * 70)

print('Numerical feature summary:')
print(df[numeric_features].describe().T.to_string())

print('Categorical feature summary:')
print(df[categorical_features].describe().T.to_string())

5. SUMMARY STATISTICS
Numerical feature summary:
                           count          mean          std     min      25%      50%       75%      max
Age                       1470.0     36.923810     9.135373    18.0    30.00     36.0     43.00     60.0
DailyRate                 1470.0    802.485714   403.509100   102.0   465.00    802.0   1157.00   1499.0
DistanceFromHome          1470.0      9.192517     8.106864     1.0     2.00      7.0     14.00     29.0
Education                 1470.0      2.912925     1.024165     1.0     2.00      3.0      4.00      5.0
EmployeeCount             1470.0      1.000000     0.000000     1.0     1.00      1.0      1.00      1.0
EmployeeNumber            1470.0   1024.865306   602.024335     1.0   491.25   1020.5   1555.75   2068.0
EnvironmentSatisfaction   1470.0      2.721769     1.093082     1.0     2.00      3.0      4.00      4.0
HourlyRate                1470.0     65.891156    20.329428    30.0    48.00     66.0     83.75    100.0
JobInv

In [7]:
# 6. Attrition-focused exploratory analysis
print('=' * 70)
print('6. ATTRITION-FOCUSED EXPLORATION')
print('=' * 70)

# These variables are selected for interpretation because they describe employee
# demographics, compensation, tenure, satisfaction, and work conditions.
selected_numeric_features = [
    'Age', 'MonthlyIncome', 'DistanceFromHome', 'TotalWorkingYears',
    'YearsAtCompany', 'JobSatisfaction', 'EnvironmentSatisfaction',
    'StockOptionLevel'
]

selected_numeric_features = [
    column for column in selected_numeric_features if column in df.columns
]

numerical_attrition_summary = (
    df.groupby(TARGET_COLUMN)[selected_numeric_features]
    .mean()
    .round(2)
    .T
)

print('Average numerical values by attrition class:')
print(numerical_attrition_summary.to_string())

# Compare attrition percentages for employees who work overtime and those who do not.
overtime_attrition_percentages = pd.crosstab(
    df['OverTime'],
    df[TARGET_COLUMN],
    normalize='index'
).mul(100).round(2)

overtime_employee_counts = pd.crosstab(df['OverTime'], df[TARGET_COLUMN])

print('Attrition percentage within each overtime group:')
print(overtime_attrition_percentages.to_string())

print('Employee count within each overtime group:')
print(overtime_employee_counts.to_string())

6. ATTRITION-FOCUSED EXPLORATION
Average numerical values by attrition class:
Attrition                     No      Yes
Age                        37.56    33.61
MonthlyIncome            6832.74  4787.09
DistanceFromHome            8.92    10.63
TotalWorkingYears          11.86     8.24
YearsAtCompany              7.37     5.13
JobSatisfaction             2.78     2.47
EnvironmentSatisfaction     2.77     2.46
StockOptionLevel            0.85     0.53
Attrition percentage within each overtime group:
Attrition     No    Yes
OverTime               
No         89.56  10.44
Yes        69.47  30.53
Employee count within each overtime group:
Attrition   No  Yes
OverTime           
No         944  110
Yes        289  127


In [8]:
# 7. Additional structural observations
print('=' * 70)
print('7. CONSTANT-VALUE COLUMNS')
print('=' * 70)

constant_columns = df.columns[df.nunique(dropna=False) == 1].tolist()

if constant_columns:
    print(f'Constant-value columns ({len(constant_columns)}):')
    for column in constant_columns:
        print(f'- {column}: {df[column].iloc[0]}')
else:
    print('No constant-value columns were found.')

7. CONSTANT-VALUE COLUMNS
Constant-value columns (3):
- EmployeeCount: 1
- Over18: Y
- StandardHours: 80


### Observations

The next cell prints data-backed observations. Values are calculated from the CSV when the notebook runs, so the conclusions are based on the provided dataset rather than assumed statistics.

In [9]:
# 8. Automatically generated, data-backed observations
print('=' * 70)
print('8. DATA-BACKED OBSERVATIONS')
print('=' * 70)

majority_class = target_counts.idxmax()
minority_class = target_counts.idxmin()

print(
    f'1. The dataset contains {df.shape[0]} employees and '
    f'{df.shape[1] - 1} predictor features.'
)

print(
    f"2. The target is imbalanced: '{majority_class}' represents "
    f'{target_percentages[majority_class]:.2f}% of employees, while '
    f"'{minority_class}' represents {target_percentages[minority_class]:.2f}%."
)

print(
    f'3. The dataset contains {len(numeric_features)} numerical and '
    f'{len(categorical_features)} categorical predictor features.'
)

print(
    f'4. The dataset has {df.isna().sum().sum()} missing values and '
    f'{duplicate_rows} fully duplicated rows.'
)

print(
    f'5. Constant-value columns found: {constant_columns}. '
    'They do not vary between employees.'
)

if {'Yes', 'No'}.issubset(numerical_attrition_summary.columns):
    print(
        f"6. Employees who left had an average MonthlyIncome of "
        f"{numerical_attrition_summary.loc['MonthlyIncome', 'Yes']:.2f}, compared with "
        f"{numerical_attrition_summary.loc['MonthlyIncome', 'No']:.2f} for employees "
        'who did not leave. This is an association, not evidence of causation.'
    )

if 'Yes' in overtime_attrition_percentages.columns:
    highest_overtime_group = overtime_attrition_percentages['Yes'].idxmax()
    highest_overtime_rate = overtime_attrition_percentages.loc[
        highest_overtime_group, 'Yes'
    ]
    print(
        f"7. The '{highest_overtime_group}' overtime group has the higher "
        f'attrition rate at {highest_overtime_rate:.2f}%.'
    )

8. DATA-BACKED OBSERVATIONS
1. The dataset contains 1470 employees and 34 predictor features.
2. The target is imbalanced: 'No' represents 83.88% of employees, while 'Yes' represents 16.12%.
3. The dataset contains 26 numerical and 8 categorical predictor features.
4. The dataset has 0 missing values and 0 fully duplicated rows.
5. Constant-value columns found: ['EmployeeCount', 'Over18', 'StandardHours']. They do not vary between employees.
6. Employees who left had an average MonthlyIncome of 4787.09, compared with 6832.74 for employees who did not leave. This is an association, not evidence of causation.
7. The 'Yes' overtime group has the higher attrition rate at 30.53%.


### Task 1 conclusion

Task 1 is complete. The next task will perform preprocessing, including justified handling of non-predictive columns, categorical variables, feature scaling where appropriate, and a reproducible train-test split.

# Task 2 - Data Preprocessing

This section prepares the Task 1 dataset for later model development. No model is trained here. All transformations are fitted using the training set only, which prevents information from the test set leaking into preprocessing.

### Preprocessing decisions

- **Missing values:** Task 1 found no missing values, so no imputation is required. This is verified again below.
- **Dropped columns:** `EmployeeCount`, `Over18`, and `StandardHours` are constant for every employee and cannot distinguish between attrition outcomes. `EmployeeNumber` is a unique employee identifier rather than an employee characteristic, so it is also excluded.
- **Target encoding:** `Attrition` is converted from `No`/`Yes` to `0`/`1`; the original target labels are not altered in `df`.
- **Categorical encoding:** One-hot encoding is used for nominal categories because their labels have no numeric order. `handle_unknown='ignore'` safely handles a category that may appear later but was not present during training.
- **Feature scaling:** Standard scaling is applied only to numerical predictors. It puts variables measured on different scales onto a comparable scale for scale-sensitive models such as Logistic Regression. One-hot encoded columns are not scaled.
- **Train-test split:** An 80/20 stratified split is used with `random_state=42`. Stratification retains the minority attrition proportion in both sets.

In [10]:
# Import only the preprocessing tools needed for Task 2.
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler

RANDOM_STATE = 42
TEST_SIZE = 0.20

print('Task 2 configuration:')
print(f'Random state: {RANDOM_STATE}')
print(f'Test-set proportion: {TEST_SIZE:.0%}')
print(f'Training-set proportion: {1 - TEST_SIZE:.0%}')

Task 2 configuration:
Random state: 42
Test-set proportion: 20%
Training-set proportion: 80%


In [11]:
# 1. Reconfirm missing values before preprocessing.
print('=' * 70)
print('TASK 2.1 - MISSING-VALUE VERIFICATION')
print('=' * 70)

missing_value_counts = df.isna().sum()
print(f'Total missing values in the dataset: {missing_value_counts.sum()}')
print('Missing-value counts by column:')
print(missing_value_counts.to_string())

if missing_value_counts.sum() == 0:
    print('Decision: No imputation is performed because the dataset has no missing values.')
else:
    raise ValueError('Missing values were found. Review the data before continuing.')

TASK 2.1 - MISSING-VALUE VERIFICATION
Total missing values in the dataset: 0
Missing-value counts by column:
Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear

In [13]:
# 2. Remove columns that cannot provide useful predictive information.
print('=' * 70)
print('TASK 2.2 - DROP CONSTANT AND IDENTIFIER COLUMNS')
print('=' * 70)

constant_columns_to_drop = ['EmployeeCount', 'Over18', 'StandardHours']
identifier_columns_to_drop = ['EmployeeNumber']
columns_to_drop = constant_columns_to_drop + identifier_columns_to_drop

missing_expected_columns = [
    column for column in columns_to_drop if column not in df.columns
]
if missing_expected_columns:
    raise ValueError(f'Expected columns were not found: {missing_expected_columns}')

print('Constant columns removed because they have one value for every employee:')
for column in constant_columns_to_drop:
    print(f'- {column}: unique values = {df[column].nunique(dropna=False)}')

print('Identifier columns removed because they identify rows rather than employee attributes:')
for column in identifier_columns_to_drop:
    print(f'- {column}: unique values = {df[column].nunique(dropna=False)}')

X = df.drop(columns=[TARGET_COLUMN] + columns_to_drop).copy()
print(f'Feature matrix shape after dropping columns: {X.shape}')
print(f'Remaining predictor columns ({X.shape[1]}):')
print(X.columns.tolist())

TASK 2.2 - DROP CONSTANT AND IDENTIFIER COLUMNS
Constant columns removed because they have one value for every employee:
- EmployeeCount: unique values = 1
- Over18: unique values = 1
- StandardHours: unique values = 1
Identifier columns removed because they identify rows rather than employee attributes:
- EmployeeNumber: unique values = 1470
Feature matrix shape after dropping columns: (1470, 30)
Remaining predictor columns (30):
['Age', 'BusinessTravel', 'DailyRate', 'Department', 'DistanceFromHome', 'Education', 'EducationField', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction', 'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked', 'OverTime', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']


In [14]:
# 3. Encode the binary target without modifying the original df column.
print('=' * 70)
print('TASK 2.3 - TARGET ENCODING')
print('=' * 70)

target_mapping = {'No': 0, 'Yes': 1}
observed_target_values = set(df[TARGET_COLUMN].dropna().unique())
unexpected_target_values = observed_target_values - set(target_mapping)

if unexpected_target_values:
    raise ValueError(f'Unexpected target values found: {unexpected_target_values}')

y = df[TARGET_COLUMN].map(target_mapping)

print(f'Target mapping: {target_mapping}')
print(f'Target vector shape: {y.shape}')
print('Encoded target value counts:')
print(y.value_counts().sort_index().to_string())

if y.isna().any():
    raise ValueError('Target encoding produced missing values.')

TASK 2.3 - TARGET ENCODING
Target mapping: {'No': 0, 'Yes': 1}
Target vector shape: (1470,)
Encoded target value counts:
Attrition
0    1233
1     237


In [15]:
# 4. Split the raw data before fitting encoders or scalers.
# Stratification preserves the attrition class balance in both partitions.
print('=' * 70)
print('TASK 2.4 - STRATIFIED TRAIN-TEST SPLIT')
print('=' * 70)

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=TEST_SIZE,
    random_state=RANDOM_STATE,
    stratify=y
)

print(f'Raw X_train shape: {X_train.shape}')
print(f'Raw X_test shape: {X_test.shape}')
print(f'y_train shape: {y_train.shape}')
print(f'y_test shape: {y_test.shape}')

print('\ny_train class counts:')
print(y_train.value_counts().sort_index().to_string())
print('y_train class percentages:')
print((y_train.value_counts(normalize=True).sort_index() * 100).round(2).to_string())

print('\ny_test class counts:')
print(y_test.value_counts().sort_index().to_string())
print('y_test class percentages:')
print((y_test.value_counts(normalize=True).sort_index() * 100).round(2).to_string())

TASK 2.4 - STRATIFIED TRAIN-TEST SPLIT
Raw X_train shape: (1176, 30)
Raw X_test shape: (294, 30)
y_train shape: (1176,)
y_test shape: (294,)

y_train class counts:
Attrition
0    986
1    190
y_train class percentages:
Attrition
0    83.84
1    16.16

y_test class counts:
Attrition
0    247
1     47
y_test class percentages:
Attrition
0    84.01
1    15.99


In [16]:
# 5. Identify categorical and numerical features from the training predictors.
print('=' * 70)
print('TASK 2.5 - FEATURE GROUPS FOR TRANSFORMATION')
print('=' * 70)

categorical_features_for_encoding = X_train.select_dtypes(
    include=['object', 'string', 'category', 'bool']
).columns.tolist()

numerical_features_for_scaling = X_train.select_dtypes(
    include='number'
).columns.tolist()

print(f'Categorical features to one-hot encode ({len(categorical_features_for_encoding)}):')
print(categorical_features_for_encoding)
print(f'\nNumerical features to standard-scale ({len(numerical_features_for_scaling)}):')
print(numerical_features_for_scaling)

if not categorical_features_for_encoding or not numerical_features_for_scaling:
    raise ValueError('Both categorical and numerical feature groups are expected.')

TASK 2.5 - FEATURE GROUPS FOR TRANSFORMATION
Categorical features to one-hot encode (7):
['BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus', 'OverTime']

Numerical features to standard-scale (23):
['Age', 'DailyRate', 'DistanceFromHome', 'Education', 'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']


In [18]:
# 6. Encode categorical variables and scale numerical variables.
# fit_transform is used only on X_train to avoid data leakage.
print('=' * 70)
print('TASK 2.6 - ENCODING AND SCALING')
print('=' * 70)

preprocessor = ColumnTransformer(
    transformers=[
        (
            'categorical_encoder',
            OneHotEncoder(handle_unknown='ignore', sparse_output=False),
            categorical_features_for_encoding
        ),
        (
            'numerical_scaler',
            StandardScaler(),
            numerical_features_for_scaling
        )
    ],
    remainder='drop',
    verbose_feature_names_out=False
)

X_train_processed_array = preprocessor.fit_transform(X_train)
X_test_processed_array = preprocessor.transform(X_test)
processed_feature_names = preprocessor.get_feature_names_out()

X_train_processed = pd.DataFrame(
    X_train_processed_array,
    columns=processed_feature_names,
    index=X_train.index
)
X_test_processed = pd.DataFrame(
    X_test_processed_array,
    columns=processed_feature_names,
    index=X_test.index
)

print(f'Number of processed features: {len(processed_feature_names)}')
print(f'Processed X_train shape: {X_train_processed.shape}')
print(f'Processed X_test shape: {X_test_processed.shape}')
print(f'Missing values in processed X_train: {X_train_processed.isna().sum().sum()}')
print(f'Missing values in processed X_test: {X_test_processed.isna().sum().sum()}')
print('\nProcessed feature names:')
print(processed_feature_names.tolist())

TASK 2.6 - ENCODING AND SCALING
Number of processed features: 51
Processed X_train shape: (1176, 51)
Processed X_test shape: (294, 51)
Missing values in processed X_train: 0
Missing values in processed X_test: 0

Processed feature names:
['BusinessTravel_Non-Travel', 'BusinessTravel_Travel_Frequently', 'BusinessTravel_Travel_Rarely', 'Department_Human Resources', 'Department_Research & Development', 'Department_Sales', 'EducationField_Human Resources', 'EducationField_Life Sciences', 'EducationField_Marketing', 'EducationField_Medical', 'EducationField_Other', 'EducationField_Technical Degree', 'Gender_Female', 'Gender_Male', 'JobRole_Healthcare Representative', 'JobRole_Human Resources', 'JobRole_Laboratory Technician', 'JobRole_Manager', 'JobRole_Manufacturing Director', 'JobRole_Research Director', 'JobRole_Research Scientist', 'JobRole_Sales Executive', 'JobRole_Sales Representative', 'MaritalStatus_Divorced', 'MaritalStatus_Married', 'MaritalStatus_Single', 'OverTime_No', 'OverTim

In [19]:
# 7. Verify scaling and final Task 2 outputs.
print('=' * 70)
print('TASK 2.7 - FINAL PREPROCESSING VERIFICATION')
print('=' * 70)

scaling_verification = pd.DataFrame({
    'Training Mean': X_train_processed[numerical_features_for_scaling].mean(),
    'Training Standard Deviation': X_train_processed[numerical_features_for_scaling].std()
}).round(4)

print('Scaled numerical-feature verification (training data only):')
print(scaling_verification.to_string())

print('\nFinal shapes after preprocessing:')
print(f'X_train_processed: {X_train_processed.shape}')
print(f'X_test_processed: {X_test_processed.shape}')
print(f'y_train: {y_train.shape}')
print(f'y_test: {y_test.shape}')

print('\nFinal class distribution in the training set:')
print(pd.DataFrame({
    'Count': y_train.value_counts().sort_index(),
    'Percentage (%)': (y_train.value_counts(normalize=True).sort_index() * 100).round(2)
}).to_string())

print('\nFinal class distribution in the test set:')
print(pd.DataFrame({
    'Count': y_test.value_counts().sort_index(),
    'Percentage (%)': (y_test.value_counts(normalize=True).sort_index() * 100).round(2)
}).to_string())

print('\nTask 2 complete. No model training has been performed.')

TASK 2.7 - FINAL PREPROCESSING VERIFICATION
Scaled numerical-feature verification (training data only):
                          Training Mean  Training Standard Deviation
Age                                -0.0                       1.0004
DailyRate                           0.0                       1.0004
DistanceFromHome                   -0.0                       1.0004
Education                          -0.0                       1.0004
EnvironmentSatisfaction             0.0                       1.0004
HourlyRate                          0.0                       1.0004
JobInvolvement                      0.0                       1.0004
JobLevel                            0.0                       1.0004
JobSatisfaction                     0.0                       1.0004
MonthlyIncome                      -0.0                       1.0004
MonthlyRate                         0.0                       1.0004
NumCompaniesWorked                  0.0                       1.0004

### Task 2 conclusion

The dataset is now split into stratified training and test sets. The fitted `preprocessor` and the processed feature matrices (`X_train_processed` and `X_test_processed`) are ready for Task 3 model development.

# Task 3 - Model Development

This section trains four classification models using the processed training data created in Task 2. It generates and stores test-set class predictions only. Model evaluation is intentionally deferred to Task 4.

### Model and hyperparameter choices

1. **Logistic Regression:** `solver='lbfgs'`, `max_iter=1000`, and `random_state=42`. The `lbfgs` solver is suitable for binary classification, while 1,000 iterations gives the optimiser sufficient time to converge on the processed features.
2. **Decision Tree:** `criterion='gini'`, `max_depth=5`, `min_samples_split=10`, `min_samples_leaf=5`, and `random_state=42`. The depth and minimum-sample limits constrain tree complexity and reduce the chance of learning overly specific training patterns.
3. **Random Forest:** `n_estimators=300`, `criterion='gini'`, `max_depth=10`, `min_samples_split=10`, `min_samples_leaf=4`, `class_weight='balanced'`, `random_state=42`, and `n_jobs=-1`. A larger collection of constrained trees provides a more stable ensemble, while balanced class weights account for the lower number of attrition cases in the training set.
4. **Bagging Classifier:** 100 decision-tree estimators, each with `max_depth=5`, `min_samples_split=10`, `min_samples_leaf=5`, and `random_state=42`; the bagging classifier uses `max_samples=1.0`, `max_features=1.0`, `random_state=42`, and `n_jobs=-1`. Bagging combines bootstrap-trained trees to reduce the instability of one decision tree.

In [20]:
# Import only the four permitted model classes used in Task 3.
from sklearn.ensemble import BaggingClassifier, RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier

print('=' * 70)
print('TASK 3 - MODEL DEVELOPMENT SETUP')
print('=' * 70)

required_task2_objects = [
    'X_train_processed', 'X_test_processed', 'y_train', 'y_test', 'RANDOM_STATE'
]
missing_task2_objects = [
    object_name for object_name in required_task2_objects
    if object_name not in globals()
]

if missing_task2_objects:
    raise RuntimeError(
        'Run all Task 1 and Task 2 cells before Task 3. Missing objects: '
        f'{missing_task2_objects}'
    )

if RANDOM_STATE != 42:
    raise ValueError(f'Expected RANDOM_STATE to be 42, found {RANDOM_STATE}.')

print(f'X_train_processed shape: {X_train_processed.shape}')
print(f'X_test_processed shape: {X_test_processed.shape}')
print(f'y_train shape: {y_train.shape}')
print(f'y_test shape: {y_test.shape}')
print('\nTraining target class counts:')
print(y_train.value_counts().sort_index().to_string())
print('\nTest target class counts:')
print(y_test.value_counts().sort_index().to_string())

TASK 3 - MODEL DEVELOPMENT SETUP
X_train_processed shape: (1176, 51)
X_test_processed shape: (294, 51)
y_train shape: (1176,)
y_test shape: (294,)

Training target class counts:
Attrition
0    986
1    190

Test target class counts:
Attrition
0    247
1     47


### Model 1 - Logistic Regression

Logistic Regression is a linear binary classifier. The numerical features were already standard-scaled in Task 2, which is appropriate because this model is sensitive to feature scale.

In [22]:
# Model 1: Logistic Regression
logistic_regression_model = LogisticRegression(
    solver='lbfgs',
    max_iter=1000,
    random_state=RANDOM_STATE
)

print('Logistic Regression hyperparameters:')
print(logistic_regression_model.get_params())

logistic_regression_model.fit(X_train_processed, y_train)
logistic_regression_predictions = logistic_regression_model.predict(X_test_processed)

print('Logistic Regression has been fitted successfully.')
print('Logistic Regression predictions have been generated.')
print(f'Prediction shape: {logistic_regression_predictions.shape}')
print('Predicted class counts (not evaluation metrics):')
print(pd.Series(logistic_regression_predictions).value_counts().sort_index().to_string())

Logistic Regression hyperparameters:
{'C': 1.0, 'class_weight': None, 'dual': False, 'fit_intercept': True, 'intercept_scaling': 1, 'l1_ratio': 0.0, 'max_iter': 1000, 'n_jobs': None, 'penalty': 'deprecated', 'random_state': 42, 'solver': 'lbfgs', 'tol': 0.0001, 'verbose': 0, 'warm_start': False}
Logistic Regression has been fitted successfully.
Logistic Regression predictions have been generated.
Prediction shape: (294,)
Predicted class counts (not evaluation metrics):
0    268
1     26


### Model 2 - Decision Tree

A Decision Tree learns rule-based splits. Its depth and minimum-sample settings are deliberately limited so that the individual tree remains easier to interpret and less likely to memorise small groups of employees.

In [23]:
# Model 2: Decision Tree
decision_tree_model = DecisionTreeClassifier(
    criterion='gini',
    max_depth=5,
    min_samples_split=10,
    min_samples_leaf=5,
    random_state=RANDOM_STATE
)

print('Decision Tree hyperparameters:')
print(decision_tree_model.get_params())

decision_tree_model.fit(X_train_processed, y_train)
decision_tree_predictions = decision_tree_model.predict(X_test_processed)

print('Decision Tree has been fitted successfully.')
print('Decision Tree predictions have been generated.')
print(f'Prediction shape: {decision_tree_predictions.shape}')
print('Predicted class counts (not evaluation metrics):')
print(pd.Series(decision_tree_predictions).value_counts().sort_index().to_string())

Decision Tree hyperparameters:
{'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'gini', 'max_depth': 5, 'max_features': None, 'max_leaf_nodes': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 5, 'min_samples_split': 10, 'min_weight_fraction_leaf': 0.0, 'monotonic_cst': None, 'random_state': 42, 'splitter': 'best'}
Decision Tree has been fitted successfully.
Decision Tree predictions have been generated.
Prediction shape: (294,)
Predicted class counts (not evaluation metrics):
0    275
1     19


### Model 3 - Random Forest

A Random Forest averages predictions from many decision trees. Class weights are balanced because attrition cases are the smaller class in the Task 2 training data.

In [24]:
# Model 3: Random Forest
random_forest_model = RandomForestClassifier(
    n_estimators=300,
    criterion='gini',
    max_depth=10,
    min_samples_split=10,
    min_samples_leaf=4,
    class_weight='balanced',
    random_state=RANDOM_STATE,
    n_jobs=-1
)

print('Random Forest hyperparameters:')
print(random_forest_model.get_params())

random_forest_model.fit(X_train_processed, y_train)
random_forest_predictions = random_forest_model.predict(X_test_processed)

print('Random Forest has been fitted successfully.')
print('Random Forest predictions have been generated.')
print(f'Prediction shape: {random_forest_predictions.shape}')
print('Predicted class counts (not evaluation metrics):')
print(pd.Series(random_forest_predictions).value_counts().sort_index().to_string())

Random Forest hyperparameters:
{'bootstrap': True, 'ccp_alpha': 0.0, 'class_weight': 'balanced', 'criterion': 'gini', 'max_depth': 10, 'max_features': 'sqrt', 'max_leaf_nodes': None, 'max_samples': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 4, 'min_samples_split': 10, 'min_weight_fraction_leaf': 0.0, 'monotonic_cst': None, 'n_estimators': 300, 'n_jobs': -1, 'oob_score': False, 'random_state': 42, 'verbose': 0, 'warm_start': False}
Random Forest has been fitted successfully.
Random Forest predictions have been generated.
Prediction shape: (294,)
Predicted class counts (not evaluation metrics):
0    259
1     35


### Model 4 - Bagging Classifier

Bagging trains several decision trees on bootstrap samples and combines their predictions. The base trees use the same constraints as the standalone tree so that the ensemble does not rely on highly complex individual trees.

In [25]:
# Model 4: Bagging Classifier with constrained decision trees.
bagging_base_estimator = DecisionTreeClassifier(
    criterion='gini',
    max_depth=5,
    min_samples_split=10,
    min_samples_leaf=5,
    random_state=RANDOM_STATE
)

bagging_classifier_model = BaggingClassifier(
    estimator=bagging_base_estimator,
    n_estimators=100,
    max_samples=1.0,
    max_features=1.0,
    random_state=RANDOM_STATE,
    n_jobs=-1
)

print('Bagging Classifier hyperparameters:')
print(bagging_classifier_model.get_params())

bagging_classifier_model.fit(X_train_processed, y_train)
bagging_classifier_predictions = bagging_classifier_model.predict(X_test_processed)

print('Bagging Classifier has been fitted successfully.')
print('Bagging Classifier predictions have been generated.')
print(f'Prediction shape: {bagging_classifier_predictions.shape}')
print('Predicted class counts (not evaluation metrics):')
print(pd.Series(bagging_classifier_predictions).value_counts().sort_index().to_string())

Bagging Classifier hyperparameters:
{'bootstrap': True, 'bootstrap_features': False, 'estimator__ccp_alpha': 0.0, 'estimator__class_weight': None, 'estimator__criterion': 'gini', 'estimator__max_depth': 5, 'estimator__max_features': None, 'estimator__max_leaf_nodes': None, 'estimator__min_impurity_decrease': 0.0, 'estimator__min_samples_leaf': 5, 'estimator__min_samples_split': 10, 'estimator__min_weight_fraction_leaf': 0.0, 'estimator__monotonic_cst': None, 'estimator__random_state': 42, 'estimator__splitter': 'best', 'estimator': DecisionTreeClassifier(max_depth=5, min_samples_leaf=5, min_samples_split=10,
                       random_state=42), 'max_features': 1.0, 'max_samples': 1.0, 'n_estimators': 100, 'n_jobs': -1, 'oob_score': False, 'random_state': 42, 'verbose': 0, 'warm_start': False}
Bagging Classifier has been fitted successfully.
Bagging Classifier predictions have been generated.
Prediction shape: (294,)
Predicted class counts (not evaluation metrics):
0    279
1     15

### Task 3 conclusion

Four fitted models and their test-set class predictions are now stored for later comparison. No performance metric or confusion matrix has been calculated in this task.

# Task 4 - Model Evaluation

This section evaluates the four Task 3 models on the same processed test set. The positive class is `1`, representing employee attrition (`Yes`).

### Evaluation metrics

- **Accuracy:** proportion of all test predictions that are correct.
- **Precision:** among employees predicted to leave, the proportion who actually leave.
- **Recall:** among employees who actually leave, the proportion correctly identified.
- **F1-score:** harmonic mean of precision and recall.
- **ROC-AUC:** ability to rank attrition cases above non-attrition cases across classification thresholds; it is calculated from probability scores, not class labels.
- **Confusion matrix:** counts true negatives, false positives, false negatives, and true positives.

In [27]:
# Import only the evaluation functions required for Task 4.
from sklearn.metrics import (
    accuracy_score,
    confusion_matrix,
    f1_score,
    precision_score,
    recall_score,
    roc_auc_score
)

print('=' * 70)
print('TASK 4 - EVALUATION SETUP')
print('=' * 70)

required_task3_objects = [
    'logistic_regression_model', 'logistic_regression_predictions',
    'decision_tree_model', 'decision_tree_predictions',
    'random_forest_model', 'random_forest_predictions',
    'bagging_classifier_model', 'bagging_classifier_predictions',
    'X_test_processed', 'y_test', 'RANDOM_STATE'
]
missing_task3_objects = [
    object_name for object_name in required_task3_objects
    if object_name not in globals()
]

if missing_task3_objects:
    raise RuntimeError(
        'Run all Task 1, Task 2, and Task 3 cells before Task 4. '
        f'Missing objects: {missing_task3_objects}'
    )

if RANDOM_STATE != 42:
    raise ValueError(f'Expected RANDOM_STATE to be 42, found {RANDOM_STATE}.')

print(f'X_test_processed shape: {X_test_processed.shape}')
print(f'y_test shape: {y_test.shape}')
print('Test-set class counts:')
print(y_test.value_counts().sort_index().to_string())

TASK 4 - EVALUATION SETUP
X_test_processed shape: (294, 51)
y_test shape: (294,)
Test-set class counts:
Attrition
0    247
1     47


In [28]:
# Generate probability scores from the already fitted Task 3 models.
# The probability for class 1 is required for correct ROC-AUC calculation.
print('=' * 70)
print('TASK 4.1 - PROBABILITY SCORE GENERATION')
print('=' * 70)

def get_positive_class_probabilities(fitted_model, features):
    """Return probabilities for the positive attrition class (1)."""
    if 1 not in fitted_model.classes_:
        raise ValueError('The fitted model does not contain positive class 1.')

    positive_class_index = list(fitted_model.classes_).index(1)
    return fitted_model.predict_proba(features)[:, positive_class_index]

logistic_regression_probability_scores = get_positive_class_probabilities(
    logistic_regression_model, X_test_processed
)
decision_tree_probability_scores = get_positive_class_probabilities(
    decision_tree_model, X_test_processed
)
random_forest_probability_scores = get_positive_class_probabilities(
    random_forest_model, X_test_processed
)
bagging_classifier_probability_scores = get_positive_class_probabilities(
    bagging_classifier_model, X_test_processed
)

probability_score_sets = {
    'Logistic Regression': logistic_regression_probability_scores,
    'Decision Tree': decision_tree_probability_scores,
    'Random Forest': random_forest_probability_scores,
    'Bagging Classifier': bagging_classifier_probability_scores
}

for model_name, probability_scores in probability_score_sets.items():
    if len(probability_scores) != len(y_test):
        raise ValueError(
            f'{model_name} probability scores do not match the test-set length.'
        )
    print(f'{model_name}: probability scores generated with shape {probability_scores.shape}.')

TASK 4.1 - PROBABILITY SCORE GENERATION
Logistic Regression: probability scores generated with shape (294,).
Decision Tree: probability scores generated with shape (294,).
Random Forest: probability scores generated with shape (294,).
Bagging Classifier: probability scores generated with shape (294,).


In [29]:
# Calculate the required evaluation metrics from Task 3 predictions and Task 4 probabilities.
print('=' * 70)
print('TASK 4.2 - METRICS AND CONFUSION MATRICES')
print('=' * 70)

model_evaluation_inputs = {
    'Logistic Regression': {
        'predictions': logistic_regression_predictions,
        'probabilities': logistic_regression_probability_scores
    },
    'Decision Tree': {
        'predictions': decision_tree_predictions,
        'probabilities': decision_tree_probability_scores
    },
    'Random Forest': {
        'predictions': random_forest_predictions,
        'probabilities': random_forest_probability_scores
    },
    'Bagging Classifier': {
        'predictions': bagging_classifier_predictions,
        'probabilities': bagging_classifier_probability_scores
    }
}

evaluation_results = []
confusion_matrices = {}

for model_name, model_outputs in model_evaluation_inputs.items():
    predictions = model_outputs['predictions']
    probability_scores = model_outputs['probabilities']

    if len(predictions) != len(y_test):
        raise ValueError(f'{model_name} predictions do not match the test-set length.')

    model_confusion_matrix = confusion_matrix(y_test, predictions, labels=[0, 1])
    confusion_matrices[model_name] = model_confusion_matrix

    evaluation_results.append({
        'Model': model_name,
        'Accuracy': accuracy_score(y_test, predictions),
        'Precision': precision_score(y_test, predictions, pos_label=1, zero_division=0),
        'Recall': recall_score(y_test, predictions, pos_label=1, zero_division=0),
        'F1-score': f1_score(y_test, predictions, pos_label=1, zero_division=0),
        'ROC-AUC': roc_auc_score(y_test, probability_scores)
    })

    print(f'{model_name}: required metrics and confusion matrix calculated.')

comparison_table = (
    pd.DataFrame(evaluation_results)
    .set_index('Model')
    .sort_values(by='F1-score', ascending=False)
    .round(4)
)

print('\nModel comparison table:')
print(comparison_table.to_string())

for model_name, model_confusion_matrix in confusion_matrices.items():
    confusion_matrix_table = pd.DataFrame(
        model_confusion_matrix,
        index=['Actual No (0)', 'Actual Yes (1)'],
        columns=['Predicted No (0)', 'Predicted Yes (1)']
    )
    print(f'\nConfusion matrix - {model_name}:')
    print(confusion_matrix_table.to_string())

TASK 4.2 - METRICS AND CONFUSION MATRICES
Logistic Regression: required metrics and confusion matrix calculated.
Decision Tree: required metrics and confusion matrix calculated.
Random Forest: required metrics and confusion matrix calculated.
Bagging Classifier: required metrics and confusion matrix calculated.

Model comparison table:
                     Accuracy  Precision  Recall  F1-score  ROC-AUC
Model                                                              
Logistic Regression    0.8605     0.6154  0.3404    0.4384   0.8115
Random Forest          0.8367     0.4857  0.3617    0.4146   0.7977
Decision Tree          0.8367     0.4737  0.1915    0.2727   0.6824
Bagging Classifier     0.8367     0.4667  0.1489    0.2258   0.7553

Confusion matrix - Logistic Regression:
                Predicted No (0)  Predicted Yes (1)
Actual No (0)                237                 10
Actual Yes (1)                31                 16

Confusion matrix - Decision Tree:
                Predic

### Discussion of model strengths, weaknesses, and use cases

- **Logistic Regression:** Its main strength is interpretability: the model has a direct coefficient for each processed feature. It may be preferred when HR needs a clear, explainable baseline. Its weakness is that it models a linear relationship in the transformed features and may miss more complex interactions.
- **Decision Tree:** Its rule-based structure is easy to explain and visualise. However, a single tree can be sensitive to the particular training sample and can miss patterns that require multiple rules or interactions. It may be preferred when straightforward decision rules are especially valuable.
- **Random Forest:** Combining many trees can model nonlinear relationships and usually reduces the instability of a single tree. Its weaknesses are reduced interpretability and higher computational cost. It may be preferred when predictive performance is more important than explaining one simple set of rules.
- **Bagging Classifier:** Bagging reduces the variance of individual decision trees by averaging many bootstrap-trained estimators. Its weaknesses are lower interpretability than one tree and the possibility that it captures less feature diversity than a random forest. It may be preferred when a more stable tree-based model is desired while retaining a controlled base-tree design.

The table and confusion matrices above provide the evidence for comparing these trade-offs. No final model recommendation is made here; that decision belongs to Task 5.

### Task 4 conclusion

All four models have been evaluated using Accuracy, Precision, Recall, F1-score, ROC-AUC, and confusion matrices. Task 5 will use this evidence to make and justify a final recommendation.

# Task 5 - Model Selection and Reflection

## a. Recommended model

I recommend **Logistic Regression** for the HR department. This recommendation is based on the complete Task 4 evidence, rather than accuracy alone.

## b. Justification for the recommendation

Logistic Regression achieved the strongest overall result in the comparison table: **Accuracy = 0.8605**, **Precision = 0.6154**, **Recall = 0.3404**, **F1-score = 0.4384**, and **ROC-AUC = 0.8115**. Its precision, F1-score, and ROC-AUC were the highest among the four evaluated models. Therefore, when it predicts that an employee will leave, that prediction is more reliable than the corresponding positive predictions from the other models in this comparison.

The confusion matrix shows that Logistic Regression correctly identified **16 of the 47** employees who actually left, while incorrectly flagging **10** employees who did not leave. It also missed **31** employees who left. This makes its recall a practical limitation: HR would still miss many potential attrition cases if it used the default classification threshold without further adjustment. Random Forest had slightly higher recall (**0.3617** versus **0.3404**), but Logistic Regression had higher precision, F1-score, ROC-AUC, and accuracy.

Interpretability is an additional advantage. Logistic Regression produces one coefficient per processed feature, making it easier to communicate the direction and relative strength of feature associations to HR stakeholders. It is also computationally lightweight and straightforward to deploy, monitor, and retrain. Its simpler linear form is generally less prone to fitting highly specific decision rules than a single tree or a large tree ensemble, although one held-out test set alone cannot prove generalisation.

For these reasons, Logistic Regression provides the best current balance of predictive performance, explainability, generalisation-oriented simplicity, and practical deployment suitability. Before operational use, HR should review the recall limitation and decide whether identifying more potential leavers is worth accepting additional false positive retention interventions.

## c. Overfitting and underfitting discussion

The available results are test-set metrics only, so they do not by themselves confirm overfitting. Confirming overfitting would require comparing training performance with validation or cross-validation performance.

The constrained Decision Tree and Bagging Classifier show possible signs of **underfitting or overly conservative positive classification**. Their recall values are low (**0.1915** and **0.1489**, respectively), and their F1-scores are also lower (**0.2727** and **0.2258**) than those of Logistic Regression and Random Forest. The Decision Tree was limited to a maximum depth of 5 with minimum-sample restrictions, and Bagging used similarly constrained base trees. These design choices reduce variance, but they may also prevent the models from learning enough patterns associated with attrition. Low recall can also be affected by the imbalanced target and the default prediction threshold, so it should not be treated as conclusive proof of underfitting.

Random Forest was also constrained through maximum depth and minimum-sample settings. It achieved a recall of **0.3617**, slightly above Logistic Regression, but had lower accuracy, precision, F1-score, and ROC-AUC than Logistic Regression. This does not establish overfitting; instead, it shows that the added nonlinear flexibility did not produce a stronger overall test-set result in this configuration.

Logistic Regression has the best test-set F1-score and ROC-AUC in the current comparison, so there is no direct test-set indication that it is overfitting. However, its recall of **0.3404** shows that the default threshold still misses many attrition cases. Cross-validation is needed to assess whether these results generalise reliably beyond this particular split.

## d. Two possible improvements

1. **Use stratified cross-validation to tune model hyperparameters and the classification threshold.** Cross-validation would provide a more reliable estimate of generalisation than one split. Tuning the probability threshold, particularly for Logistic Regression, could improve recall if HR decides that identifying more potentially departing employees is more valuable than reducing false positives.

2. **Develop and validate additional job-related features.** Examples could include carefully justified tenure ratios or income-related features derived from available fields, and, if permitted, additional HR data that is available at prediction time. More informative predictors may help distinguish employees likely to leave without relying only on the current variables.